src/inner.jl

```diff
@@ -0,0 +1,143 @@
+using ITensors: inner, scalar, loginner
+using LinearAlgebra: norm, norm_sqr
+
+default_contract_alg(tns::Tuple) = "bp"
+
+function ITensors.inner(
+  ϕ::AbstractITensorNetwork,
+  ψ::AbstractITensorNetwork;
+  alg=default_contract_alg((ϕ, ψ)),
+  kwargs...,
+)
+  return inner(Algorithm(alg), ϕ, ψ; kwargs...)
+end
+
+function ITensors.inner(
+  ϕ::AbstractITensorNetwork,
+  A::AbstractITensorNetwork,
+  ψ::AbstractITensorNetwork;
+  alg=default_contract_alg((ϕ, A, ψ)),
+  kwargs...,
+)
+  return inner(Algorithm(alg), ϕ, A, ψ; kwargs...)
+end
+
+function ITensors.inner(
+  alg::Algorithm"exact",
+  ϕ::AbstractITensorNetwork,
+  ψ::AbstractITensorNetwork;
+  sequence=nothing,
+  contraction_sequence_kwargs=(;),
+  kwargs...,
+)
+  tn = inner_network(ϕ, ψ; kwargs...)
+  if isnothing(sequence)
+    sequence = contraction_sequence(tn; contraction_sequence_kwargs...)
+  end
+  return scalar(tn; sequence)
+end
+
+function ITensors.inner(
+  alg::Algorithm"exact",
+  ϕ::AbstractITensorNetwork,
+  A::AbstractITensorNetwork,
+  ψ::AbstractITensorNetwork;
+  sequence=nothing,
+  contraction_sequence_kwargs=(;),
+  kwargs...,
+)
+  tn = inner_network(ϕ, A, ψ; kwargs...)
+  if isnothing(sequence)
+    sequence = contraction_sequence(tn; contraction_sequence_kwargs...)
+  end
+  return scalar(tn; sequence)
+end
+
+function ITensors.loginner(
+  ϕ::AbstractITensorNetwork,
+  ψ::AbstractITensorNetwork;
+  alg=default_contract_alg((ϕ, ψ)),
+  kwargs...,
+)
+  return loginner(Algorithm(alg), ϕ, ψ; kwargs...)
+end
+
+function ITensors.loginner(
+  ϕ::AbstractITensorNetwork,
+  A::AbstractITensorNetwork,
+  ψ::AbstractITensorNetwork;
+  alg=default_contract_alg((ϕ, A, ψ)),
+  kwargs...,
+)
+  return loginner(Algorithm(alg), ϕ, A, ψ; kwargs...)
+end
+
+function ITensors.loginner(
+  alg::Algorithm"exact", ϕ::AbstractITensorNetwork, ψ::AbstractITensorNetwork; kwargs...
+)
+  return log(inner(alg, ϕ, ψ); kwargs...)
```

Comment:
Hmm this doesn't seem like the right logic. The point of `loginner` is to try to avoid overflow if `inner` is very large. Should `inner` call `loginner`?

Reply:
We haven't actually defined an algorithm for computing `loginner(alg::Algorithm"exact", ....)` yet? So this is why I put the logic this way round for now.

But I see that it kind of defeats the point to have it because `log` is to deliberately avoid over/ underflow. I could remove this function for now and not support `loginner(alg::Algorithm"exact", ....)` until we actually incorporate an algorithm for it in? Then once we do we can pass `inner()` to it and take the `exp`.

Reply:
Oh I see, sorry, I hadn't realized this was the `::Algorithm"exact"` version. I think it is ok to keep this as-is as a suboptimal stand-in.